File: day15/day15.cpp
#include <fstream>
#include <iostream>

#include <fmt/format.h>

using namespace std;

static pair<size_t, size_t> load(istream &input)
{
	size_t r[2] = {};
	for (size_t i = 0; i < 2; i++)
	{
		string line;
		getline(input, line);
		if (input.fail())
		{
			break;
		}
		size_t pos = 0;
		size_t next;
		while ((next = line.find(' ', pos)) != string::npos)
		{
			pos = next + 1;
		}
		r[i] = stoi(line.substr(pos));
	}
	return make_pair(r[0], r[1]);
}

static size_t part1(size_t a, size_t b)
{
	size_t count = 0;
	for (size_t i = 0; i < 40000000ULL; i++)
	{
		a = (a * 16807) % 2147483647ULL;
		b = (b * 48271) % 2147483647ULL;
		count += (a & 0xffff) == (b & 0xffff);
	}
	return count;
}

static size_t part2(size_t a, size_t b)
{
	size_t count = 0;
	for (size_t i = 0; i < 5000000ULL; i++)
	{
		do
		{
			a = (a * 16807) % 2147483647ULL;
		} while (a & 3);

		do
		{
			b = (b * 48271) % 2147483647ULL;
		} while (b & 7);

		count += (a & 0xffff) == (b & 0xffff);
	}
	return count;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	auto [a, b] = load(input);
	input.close();
	if (input.fail())
	{
		cerr << "Cannot parse the data" << endl;
		return 1;
	}

	cout << fmt::format("Part1: {}", part1(a, b)) << endl
	     << fmt::format("Part2: {}", part2(a, b)) << endl;
	return 0;
}

File: day15/day15.py
#!/usr/bin/env python3

def parse(txt):
    lst = txt.splitlines()
    a = int(lst[0].split("with ")[1])
    b = int(lst[1].split("with ")[1])
    return a, b

def part1(a, b):
    count = 0
    for i in range(40000000):
        a = (a * 16807) % 2147483647
        b = (b * 48271) % 2147483647
        if (a & 0xffff) == (b & 0xffff):
            count += 1
    return count

def part2(a, b):
    count = 0
    for i in range(5000000):
        while True:
            a = (a * 16807) % 2147483647
            if a & 3 == 0:
                break

        while True:
            b = (b * 48271) % 2147483647
            if b & 7 == 0:
                break

        if (a & 0xffff) == (b & 0xffff):
            count += 1
    return count

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    a, b = parse(txt)
    print("Part1:", part1(a, b))
    print("Part2:", part2(a, b))
